### ISSUE TITLE
A canceled task that completes fails with "Job was not found or not in "doing" or "todo" status"

### ISSUE THREAD

**Issue by wokis** [procrastinate-org/procrastinate#511]
A task that is canceled has it status set to `FAILED`, however when the task completes only `DOING` and `TODO` is accepted for task completion. This causes `procrastinate` to fail and terminate it's process.

https://github.com/procrastinate-org/procrastinate/blob/65829b3fc17374985ee17c2911083f60132f1bfd/procrastinate/shell.py#L142 https://github.com/procrastinate-org/procrastinate/blob/65829b3fc17374985ee17c2911083f60132f1bfd/procrastinate/sql/schema.sql#L207

```
Dec 30 16:24:00 host1 python3[1280546]: Error:
Dec 30 16:24:00 host1 python3[1280546]:     Database error.
Dec 30 16:24:00 host1 python3[1280546]:     
Dec 30 16:24:00 host1 python3[1280546]: Job was not found or not in "doing" or "todo" status (job id: 47)
Dec 30 16:24:00 host1 python3[1280546]: CONTEXT:  PL/pgSQL function procrastinate_finish_job(integer,procrastinate_job_status,timestamp with time zone,boolean) line 24 at RAISE
Dec 30 16:24:00 host1 systemd[1]: procrastinate.service: Main process exited, code=exited, status=1/FAILURE
Dec 30 16:24:00 host1 systemd[1]: procrastinate.service: Failed with result 'exit-code'.
```

**Comment by ewjoachim** [comment]
That's true.
Let me rephrase reproduction steps:
- Defer a task that takes some time to complete
- Have a worker start executing the task
- While the task executes, cancel the task using the shell (`procrastinate shell / cancel <id>`)
- When the task finishes, it will crash the worker.

I would say that the problem is that the shell should not accept to cancel a task that has started. It's confusing to call this "cancellation" given that the task will not be interrupted.

If we want to include a "real" cancellation feature at some point, that is able to interrupt the running task, then we'll need to do that. Meanwhile we should just not make it possible to cancel an already running task.
(it's fine to cancel a task before it starts though)

**Comment by wokis** [comment]
I was trying to implement task cancelation using the same principles as the shell when I ran into this bug.

It would be nice if `JobContext.job` status was updated when a job's `status` changes. Right now I'm settling for periodically checking the `procrastinate_jobs` table.

When you say we need to use `SELECT FOR UPDATE` do you mean only in the `procrastinate_finish_job` SQL function or do we need to do locking also when checking the current status of a task?

I'm willing to spend some time on this (mostly making the status of a `JobContext.job` update when a new status is set) if there is any interest in such functionality?

**Comment by ewjoachim** [comment]
> When you say we need to use SELECT FOR UPDATE do you mean only in the procrastinate_finish_job SQL function or do we need to do locking also when checking the current status of a task?

Ah I just meant when checking the state of the state before calling `procrastinate_finish_job` to ensure we don't finish a cancelled task.

> It would be nice if JobContext.job status was updated when a job's status changes. Right now I'm settling for periodically checking the procrastinate_jobs table.

It could but it would be a change of paradigm. Currently, this is just a Python attribute lookup, and I'd like to avoid having hidden I/O (simple attribute lookups that trigger database operations). It looks innocent at first, up until the point you realize it creates all kinds of funky links in your codebase, and simple things start becoming hard.

That being said, I'm absolutely not opposed to exposing functions that would give this information in a nice way. That being said, I think there's already something you can use in `procrastinate.manager.JobManager.list_jobs_async(id=<your job id>)`. I believe this will give you access to the job's state.

**Comment by wokis** [comment]
`procrastinate.manager.JobManager.list_jobs(id=<your job id>)` should also work, shouldn't it? However I can't seem to get it to work with synchronous tasks as listing the jobs raises a `RuntimeError('This event loop is already running')`.

**Comment by ewjoachim** [comment]
Ah, yeah, the problem is that the loop is not re-entrant...

The relevant code is:
https://github.com/procrastinate-org/procrastinate/blob/master/procrastinate/utils.py#L190-L199

Considering we're in a task, so in async code, calling some sync code, which then wants to call async code... Yeah, it sucks :( I don't know if the answer should be threading (:skull: :fearful:) or some kinds of shenanigans or what...

**Comment by wokis** [comment]
Had time to return to this issue. I was thinking about someting like this:
```
CREATE FUNCTION procrastinate_finish_job(job_id integer, end_status procrastinate_job_status, next_scheduled_at timestamp with time zone, delete_job boolean) RETURNS void
    LANGUAGE plpgsql
AS $$
DECLARE
    _job_id bigint;
BEGIN
    IF end_status NOT IN ('succeeded', 'failed') THEN
        RAISE 'End status should be either "succeeded" or "failed" (job id: %)', job_id;
    END IF;
    IF delete_job THEN
        DELETE FROM procrastinate_jobs
        WHERE id = job_id AND status IN ('todo', 'doing', 'failed')
        RETURNING id INTO _job_id;
    ELSE
        SELECT id INTO _job_id FROM procrastinate_jobs
        WHERE id = job_id AND status IN ('todo', 'doing', 'failed')
        FOR UPDATE;

        UPDATE procrastinate_jobs
        SET status = end_status,
            attempts =
                CASE
                    WHEN status = 'doing' THEN attempts + 1
                    ELSE attempts
                END
        WHERE id = job_id AND status IN ('todo', 'doing')
        RETURNING id INTO _job_id;
    END IF;
    IF _job_id IS NULL THEN
        RAISE 'Job was not found or not in "doing", "todo" or "failed" status (job id: %)', job_id;
    END IF;
END;
$$;
```

This would be the easiest option (the other being adding a "CANCELED" status but would require a bit more code) and then we could work on things outlined in https://github.com/procrastinate-org/procrastinate/issues/780#issuecomment-1529005989 to make it more feature complete, cause right now each job would have to resort in pulling the database at some desired rate as there is no signaling to indicate a job was canceled.

**Comment by ewjoachim** [comment]
I think it would make sense to add `CANCELED` and maybe also `ABORTED` as possible states (though `ABORTED` is not possible as of now, but once it is, it will help knowing whether part of the job ran vs none at all). I'm afraid doubling down calling canceled tasks "failed" will only lead to more ambiguity down the line. That was a bad choice I made, it's probably time to revisit it :)

**Comment by medihack** [comment]
@ewjoachim I can look into this issue as we need a similar feature. For our use case, I thought more about just deleting the revoked job from the database and only allow to revoke jobs that are in `todo` state, but I could think about a more general solution that can be integrated into Procrastinate itself.

I would still keep things simple and only allow to cancel / revoke jobs that are in a `todo` state and not those that are already in `doing` state because I can't think of a clean solution to handle canceling / aborting an already running job. The maximum I can think of is to just set a `canceled` state so that the user code can react accordingly. What is your opinion about that?

**Comment by ewjoachim** [comment]
My opinion (also partly in https://github.com/procrastinate-org/procrastinate/issues/780)

- It would be nice to have a cancellation feature
- It's going to be trivial when a job is in `todo`, and hard when it's already `doing`. Maybe we even want 2 distinct mechanisms (with a single entrypoint that could call either `def cancel(even_if_already_started: bool)` (naming TBD))
- (As you mentioned,) we probably want 3 states for representing what's happening: `canceling`, `cancelled`, `revoked` (wording TBD, I really don't know between `cancel`, `revoke`, `repeal`, `abort`, `abandon`, `scrap`, `rescind`, `withdraw` etc. what best expresses the idea of "cancelling" a task before vs after it started.) but you're right that in the case of after it started, we'll likely want to know when a task has received the order vs not yet, and when it's officially stopped
- We should probably use listen-notify for workers to be aware of cancellation request. I think it will be overkill to have a channel per task, so it's likely we'll use a single channel (possibly even the existing channel with different message types). This would mean we'd need to start actually paying attention to the content of the message. Also, we'll need the listen-notify loop to be able to tell the "worker" & "subworker" (there's an ongoing discussion on Discord on the naming of those around https://github.com/procrastinate-org/procrastinate/issues/933). Same as the listen/notify for new job: maybe we also want occasional polling so that if listen/notify doesn't work, or we somehow missed the info), the worker can still somehow know it needs to cancel the task, even if it's late.
- There's a design decision to make: do we want it to be possible for tasks to handle graceful cancellation ? I don't know. Let's assume yes below, but it's fully possible we'll decide no.
- I'd say the cancellation will happen by calling `task.cancel()` on the job coroutine, and somehow changing something withing `job_context`. I'm saying "somehow" because `job_context` currently is a frozen dataclass, so the only way we'd get something to change is by passing a mutable object. Calling `task.cancel()` will make the job raise at the next event loop tick. This will do nothing if the job's task is a synchronous function because `sync_to_async` [shields from cancellation](https://github.com/django/asgiref/blob/f86642b838b6b1c6b14ca18452d30de5069f01f4/asgiref/sync.py#L468) (as far as I can tell). We could (should) offer the ability for jobs to shield from cancellation too but it may be easily done by client code. Whether the client code is asynchronous and does something with the `CancelledError` exception or it periodically reads `job_context` to exit the loop early (which works both in synchronous, shielded asynchronous or asynchronous with a `try/except CancelledError:`), we can reasonably hope that it's going to try and end quickly.
- But a job (especially a synchronous job) might be completely oblivious that it's supposed to get cancelled, so we might want to have some other ungraceful mechanism to cancel a task maybe ? Not sure.
- Last thing, do we want to call cancel on running tasks when the worker is stopped? There could be multiple things: not taking any new tasks but letting running tasks finish, or cancelling running tasks, or killing everything. Or doing a first action, and then escalating after a given time.

There seem to be a few decisions to make, and potentially a few knobs that will be needed for the experience to be optimal for everyone. That's what I have in mind around cancellation.

**Comment by medihack** [comment]
Alright, let's figure out a more concrete plan together.

> * It would be nice to have a cancellation feature

Absolutely 😉

> * It's going to be trivial when a job is in `todo`, and hard when it's already `doing`. Maybe we even want 2 distinct mechanisms (with a single entrypoint that could call either `def cancel(even_if_already_started: bool)` (naming TBD))

How about `def cancel(abort: bool)` (see below).

> * (As you mentioned,) we probably want 3 states for representing what's happening: `canceling`, `cancelled`, `revoked` (wording TBD, I really don't know between `cancel`, `revoke`, `repeal`, `abort`, `abandon`, `scrap`, `rescind`, `withdraw` etc. what best expresses the idea of "cancelling" a task before vs after it started.) but you're right that in the case of after it started, we'll likely want to know when a task has received the order vs not yet, and when it's officially stopped

How about those 3 states: `cancelled` (or the more used American English one `canceled`), `aborting`, `aborted`. I think those are the most used ones and are [well distinguishable](https://english.stackexchange.com/a/535155):

_**Cancel** implies the action is rescinded before it implements, possibly consequence-free. It's the word used to bow out gracefully when prompted to confirm an order.
**Abort** is an emergency procedure to interrupt an action already in progress because continuing would be disastrous._

So a task that is still in `todo` state just gets `canceled`, and a task that is in `doing` state gets the `aborting` state. And if the user handles the abort request (see below), it gets the `aborted` state.

> * We should probably use listen-notify for workers to be aware of cancellation request. I think it will be overkill to have a channel per task, so it's likely we'll use a single channel (possibly even the existing channel with different message types). This would mean we'd need to start actually paying attention to the content of the message. Also, we'll need the listen-notify loop to be able to tell the "worker" & "subworker" (there's an ongoing discussion on Discord on the naming of those around [Worker refactor #933](https://github.com/procrastinate-org/procrastinate/issues/933)). Same as the listen/notify for new job: maybe we also want occasional polling so that if listen/notify doesn't work, or we somehow missed the info), the worker can still somehow know it needs to cancel the task, even if it's late.

That sounds a bit more "heavy-weight" than what I had in mind. I thought more about letting the user poll the database (by a special method in the context, e.g. `context.is_canceled()`). I wonder what applications would benefit the most from aborting an already ongoing job. In our case, there is a task that transfers radiological images in the background, which may take several minutes. It would be nice if we could abort this, but it must not be instantly. It's ok if it takes a few seconds, but it should do it gracefully. As it takes minutes, an occasional database request doesn't matter at all. But we also have other short-running tasks (that take only a few seconds or even less), where we are not interested in aborting at all (we would just let them run through even if the task is canceled). **I don't say that I am against using listen-notify**, but I am not sure if the more complicated implementation is really worth it.

> * There's a design decision to make: do we want it to be possible for tasks to handle graceful cancellation ? I don't know. Let's assume yes below, but it's fully possible we'll decide no.
> * I'd say the cancellation will happen by calling `task.cancel()` on the job coroutine, and somehow changing something withing `job_context`. I'm saying "somehow" because `job_context` currently is a frozen dataclass, so the only way we'd get something to change is by passing a mutable object. Calling `task.cancel()` will make the job raise at the next event loop tick. This will do nothing if the job's task is a synchronous function because `sync_to_async` [shields from cancellation](https://github.com/django/asgiref/blob/f86642b838b6b1c6b14ca18452d30de5069f01f4/asgiref/sync.py#L468) (as far as I can tell). We could (should) offer the ability for jobs to shield from cancellation too but it may be easily done by client code. Whether the client code is asynchronous and does something with the `CancelledError` exception or it periodically reads `job_context` to exit the loop early (which works both in synchronous, shielded asynchronous or asynchronous with a `try/except CancelledError:`), we can reasonably hope that it's going to try and end quickly.
> * But a job (especially a synchronous job) might be completely oblivious that it's supposed to get cancelled, so we might want to have some other ungraceful mechanism to cancel a task maybe ? Not sure.

In my opinion (and also as a user of Procrastinate), I really want to handle an abort request on my own. That way, it is completely in the user's hand whether he wants to do it gracefully or more abruptly. We then also don't have the problem of thinking about whether we can support every kind of task (synchronous or asynchronous). Because if the task is synchronous, I don't see an easy way to forcefully abort it (if possible at all). But the user itself can do it easily. For example, in our application, there is also one task that can't be terminated gracefully (because of some 3rd party library). We start a separate process in our (synchronous) task (using multiprocessing) and monitor the cancel state (every 10 seconds or so). If the task gets canceled, we simply terminate the previously started process. Quite harsh, but it works.

> * Last thing, do we want to call cancel on running tasks when the worker is stopped? There could be multiple things: not taking any new tasks but letting running tasks finish, or cancelling running tasks, or killing everything. Or doing a first action, and then escalating after a given time.

I would maybe handle this in another issue. Cancellation and abortion are, in my opinion, more something a user does on his own. But yes, some labeling of jobs that were in the middle of being processed while the worker was stopped would be nice 😃.

**Comment by medihack** [comment]
I gave it another thought, and here is some suggestion. I could implement it in two steps.

In the first PR, I would implement the more "lightweight" aborting feature as mentioned above (by letting the user poll the abortion state from the database with `context.is_canceled()`). I guess it is pretty easy to implement and I could get it done in a few days (we need this feature anyway for our application).

Then, in a second PR, I could implement the listen-notify feature on top of it. It will likely take a few weeks and is best done after the worker refactoring #933 is finished. It won't break anything, but `context.is_canceled()` wouldn't then poll the database anymore. Instead, its state is updated in the background by the database notification. I also could add an additional method to get a "real-time" notification when the job is aborted, e.g., `context.listen_canceled(cb: Callable)`.

I can even think of a possible third PR to forcefully terminate a task with a `CancelledError`, but I am still not a very big fan of it as it handles asynchronous tasks differently than synchronous tasks. On the other hand, it would still be doable and the other two PRs would be a good base for that.

**Comment by ewjoachim** [comment]
This starts to sound like a solid plan.

A few notes:

I'm a tiny bit worried about abort/aborting because it calls a displeasant context in mind due to the political situation & such. See e.g. [this](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/abort-abortion) or [this](https://ux.stackexchange.com/questions/75694/should-i-avoid-using-abort-in-favor-of-its-synonyms). I know it might not be a opinion everyone agrees with, I think it's worth spending an additional minute at the time no decision is made yet to avoid having to imagine changing the term later. Let's imagine `stop` for now.

If you have `job.is_canceled` (well maybe we should be using the other one: or `should_stop()` ? But that's not the point). At first this thing is going to do I/Os, so we need a sync and async version `should_stop()`, `should_stop_async()` even though `should_stop_async()` would soon do no I/O at all. It's not problematic per se, we'll need to decide if we keep it, keep it deprecated, remove it (as long as we make the API provisional, I'm ok with that). Alternatively, we could have: `should_stop(use_cache: bool=True)`

I'm a bit afraid that a large number of tasks running in parallel (thanks to async) each often calling `should_stop` before the cache is introduced, and doing it in every loop might flood the DB. That's a problem users may have everything to deal with on their own (e.g. by not checking too often).

Also: I'm trying to think about the use case we're solving and put myself in the shoes of a user who would be using this:
So either your job has a loop and you check for `should_stop()` once in a while, or you have a multi-step job and you check between step. If it's the second one, the idea that you want to stop gracefully implies that you save each step and when you re-run, you have implemented a way to skip specific steps. Not saying you can't but it sounds like an aweful lot of work, compared to having each step be its own job.

Alternatively, you have a loop. Similarly, if you want to benefit from graceful stop, it means progress is saved throughout the loop. It bears the question: why isn't each item of the loop in its own job? I only see one possibility that makes sense: the setup time (both the procrastinate job startup time and the time the user job takes to prepare for processing items) is much higher than the time it takes to process a single item.

In this case, this makes me think I understand that a task queue might not be what the user really wants. It may be a pragmatic solution to the problem, but it means they have to implement their own mini-task-queue within the task queue:
Either they submit batches of items to process, or they're submitting tasks that will keep processing items until there's no item to process, but it both cases, it looks like the fact there are procrastinate jobs with their own steps and all is a hindrance, where they'd actually want a server that just keeps processing the items.

Of course, I can imagine that the reality might be that there are a lot of varied things to do where items processing would be a small part and they wouldn't want dedicated processes for that but if we're at the level where graceful stop makes sense, the line seems thin before a dedicated process makes more sense.

What am I missing ?

**Comment by medihack** [comment]
> I'm a tiny bit worried about abort/aborting because it calls a displeasant context in mind due to the political situation & such. See e.g. [this](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/abort-abortion) or [this](https://ux.stackexchange.com/questions/75694/should-i-avoid-using-abort-in-favor-of-its-synonyms). I know it might not be a opinion everyone agrees with, I think it's worth spending an additional minute at the time no decision is made yet to avoid having to imagine changing the term later. Let's imagine `stop` for now.

I am okay with that, even when I think `abort` is the correct word in that context. It's also not visible to a _general audience_ or appears in a _public UI_. A job pipeline is visible to the developers but not to the end user (this is what the Microsoft and StackExchange article means). [It's](https://docs.celeryq.dev/en/stable/reference/celery.contrib.abortable.html#celery.contrib.abortable.AbortableAsyncResult.abort) [also](https://github.com/Flared/dramatiq-abort) [the](https://docs.rs/tokio/latest/tokio/task/struct.AbortHandle.html) [most](https://docs.sevenbridges.com/reference/abort-a-task) [used](https://docs.particular.net/nservicebus/pipeline/aborting) [term](https://www.jenkins.io/doc/book/using/aborting-a-build/) [for](https://help.salesforce.com/s/articleView?id=000385103&type=1) [job](https://www.gurobi.com/documentation/current/remoteservices/aborting_jobs.html) [pipelines](https://docs.oracle.com/cd/E50245_01/E50252/html/vmcli-ref-abort-job.html) 😉. `stop` isn't such a good word as it doesn't express that it happened right in the middle of something. If you are really against `aborted`, how about `interrupted`?

> 
> If you have `job.is_canceled` (well maybe we should be using the other one: or `should_stop()` ? But that's not the point). At first this thing is going to do I/Os, so we need a sync and async version `should_stop()`, `should_stop_async()` even though `should_stop_async()` would soon do no I/O at all. It's not problematic per se, we'll need to decide if we keep it, keep it deprecated, remove it (as long as we make the API provisional, I'm ok with that). Alternatively, we could have: `should_stop(use_cache: bool=True)`

That's a good point. We could also leave those methods there as a polling solution and introduce that listener option later for handling the notification (as mentioned in my message above, something like `context.listen_interrupt(cb: Callable)`). It's just something we have to clearly document. I still think polling is enough for most use cases (more on this below).

> 
> I'm a bit afraid that a large number of tasks running in parallel (thanks to async) each often calling `should_stop` before the cache is introduced, and doing it in every loop might flood the DB. That's a problem users may have everything to deal with on their own (e.g. by not checking too often).

Absolutely, and it should be clearly documented in one of those nice warning boxes in the documentation.

> 
> Also: I'm trying to think about the use case we're solving and put myself in the shoes of a user who would be using this: So either your job has a loop and you check for `should_stop()` once in a while, or you have a multi-step job and you check between step. If it's the second one, the idea that you want to stop gracefully implies that you save each step and when you re-run, you have implemented a way to skip specific steps. Not saying you can't but it sounds like an aweful lot of work, compared to having each step be its own job.

Interrupting something gracefully is most often challenging. That's why we can't take this away from the user. For our use cases (more on those below), I can handle the abortion/interruption on my own (by using a flag in our models). But I would like Procrastinate to help me just a bit with that.

> 
> Alternatively, you have a loop. Similarly, if you want to benefit from graceful stop, it means progress is saved throughout the loop. It bears the question: why isn't each item of the loop in its own job? I only see one possibility that makes sense: the setup time (both the procrastinate job startup time and the time the user job takes to prepare for processing items) is much higher than the time it takes to process a single item.
> 
> In this case, this makes me think I understand that a task queue might not be what the user really wants. It may be a pragmatic solution to the problem, but it means they have to implement their own mini-task-queue within the task queue: Either they submit batches of items to process, or they're submitting tasks that will keep processing items until there's no item to process, but it both cases, it looks like the fact there are procrastinate jobs with their own steps and all is a hindrance, where they'd actually want a server that just keeps processing the items.

I can speak for our applications (and we will be the users of Procrastinate 😉). Two jobs come directly to my mind where to deal with abortion/interruption.
The first one (as mentioned in a previous post) is to transfer radiological studies (CT or MRI), which consist of many individual images. However, one study has to be processed together, as those get pseudonymized, which must be done in the same process to keep things consistent. It's not possible to process each image in a single job.
Another feature we have is letting a user process radiological reports by piping them through an LLM. We have a big Nvidia cluster and can process millions of those reports per day. To do that, we have to create batches to send them in parallel to the Nvidia GPUs. Some of those batches are pooled together in a Celery task.
We are pretty happy with how things work, but I want to replace Celery with Procrastinate to simplify our infrastructure. [Celery helps us with aborting a task](https://docs.celeryq.dev/en/stable/reference/celery.contrib.abortable.html#celery.contrib.abortable.AbortableAsyncResult.abort), and this is the same thing I would like to have in Procrastinate 😊.

> 
> Of course, I can imagine that the reality might be that there are a lot of varied things to do where items processing would be a small part and they wouldn't want dedicated processes for that but if we're at the level where graceful stop makes sense, the line seems thin before a dedicated process makes more sense.

I think aborting/interrupting a job is not a feature that should often be used. Usually, jobs are processed so fast that abortion/interruption doesn't make sense. But there are scenarios (like the ones I mentioned) where it does make sense. And it would be nice if Procrastinate (like Celery) would help a tiny bit with that.

**Comment by ewjoachim** [comment]
> If you are really against aborted, how about interrupted?

I'm not **really** against `abort`, just a tiny bit worried. Interrupt is ok, but as you showed, `abort` is the usual term for this, and this adds weight to the choice (it will help adoption). Let's roll with `abort`.

> ...

So if I understand correctly (and tell me if I'm wrong), you're interested in canceling, but not specifically in graceful cancelling, right? If there was a way that the coroutine (or the underlying thread) was killed (ideally with an exception so it would execute the context exits, and the `try/finalize` blocks), this would be what you'd want right ?

But then graceful (cooperative) termination looks simple in both async and sync, ungraceful looks simple in async and much more complex in sync (because there doesn't seem to be an obvious way of interrupting a thread), so it's not much that we'd want graceful termination because it's graceful, but because ungraceful seems harder to implement. Right ?

If so, I'm good to go with this plan :)

**Comment by medihack** [comment]
> So if I understand correctly (and tell me if I'm wrong), you're interested in canceling, but not specifically in graceful cancelling, right? If there was a way that the coroutine (or the underlying thread) was killed (ideally with an exception so it would execute the context exits, and the `try/finalize` blocks), this would be what you'd want right ?
> 
> But then graceful (cooperative) termination looks simple in both async and sync, ungraceful looks simple in async and much more complex in sync (because there doesn't seem to be an obvious way of interrupting a thread), so it's not much that we'd want graceful termination because it's graceful, but because ungraceful seems harder to implement. Right ?

I can't fully follow anymore 😂. What I want to implement, at least in a first PR is very similar to [what Celery does with its `AbortableTask`](https://docs.celeryq.dev/en/2.1-archived/reference/celery.contrib.abortable.html#usage-example).

Here is some Pseudocode to make things more clear and concrete.

For example, inside some Django view
```python
# Cancel a specific job by using the job manager
app.job_manager.cancel_task_by_id(325, abort=True)
# or maybe using the task
my_task.cancel_by_id(325, abort=True)
```

If the job is still in `todo` status it just gets the `canceled` status and will no longer be processed by any worker. If the job is in `doing` status it will get the `aborting` status. This `aborting` status can be respected by the task (but it doesn't have to).

In our example task definition:
```python
@app.task(pass_context=True)
def my_task(context):
  for i in range(100):
    if context.should_abort(): # polls the current status from the database
      # When the task raises JobAborted the status will be "aborted". When it raises another exception it will be "failed".
      # If it raises nothing, it will be "succeeded". So, it's up to the task/developer to decide what is best for this specific task.
      raise exceptions.JobAborted
    do_something_expensive()
```

Of course, we also need the async counterparts for those methods, e.g. `app.job_manager.cancel_task_by_id_async`.

And as I said, this would be the first PR. A second PR could then add something even more sophisticated with listen-notify.

So, do I get your ok?

**Comment by ewjoachim** [comment]
You get my ok :)

**Comment by onlyann** [comment]
👋 

Keeping an eye on this conversation as it is related to the worker refactoring bit I started.

I wanted to share a few thoughts on the topic.

The first thing that crossed my mind when reading about cancelling and aborting is how confusing it can be when you consider the definition of task cancellation from asyncio.

Granted, this is talking about procrastine tasks but given how embedded asyncio is within the library, would it help to share that definition of cancellation?

Another thought is that it appears there are multiple features being discussed at once: 
- being able to differentiate a job that failed from a job that was cancelled
- being able to cancel a specific job remotely
- being able to react to a cancellation in user code

I am wondering if those features should be added incrementally.

Let me also expand a bit on what my intentions were in regard to cancellation as part of the worker refactoring.

My idea is to introduce a setting to be able to tell procrastinate how long it should wait for in flight jobs to complete by themselves during the graceful shutdown (the current default is forever).

Whenever it goes beyond this duration, the worker would cancel the underlying (asyncio) tasks. 

The user code could then react to it in the standard asyncio way.

Now, it is a different kettle of fish for synchronous tasks. 

On that note, it begs the question of whether procrastinate wants to keep the burden of handling sync code itself in the future. 
If the answer is no, then maybe it is good enough. Otherwise, it sounds like a harder problem to solve.

**Comment by medihack** [comment]
> Keeping an eye on this conversation as it is related to the worker refactoring bit I started.
> 
> I wanted to share a few thoughts on the topic.

Sounds good, I am always interested in other opinions and how I can improve PR #1081.

> The first thing that crossed my mind when reading about cancelling and aborting is how confusing it can be when you consider the definition of task cancellation from asyncio.
> 
> Granted, this is talking about procrastine tasks but given how embedded asyncio is within the library, would it help to share that definition of cancellation?
> 
> Another thought is that it appears there are multiple features being discussed at once:
> 
>     * being able to differentiate a job that failed from a job that was cancelled
> 
>     * being able to cancel a specific job remotely
> 
>     * being able to react to a cancellation in user code
> 
> 
> I am wondering if those features should be added incrementally.

With PR #1081, I am trying to create a solid foundation for **user-initiated** cancellation. Other PRs can build stuff on top of it. I am pretty sure that we need the following states (the naming can still be discussed):
- `cancelled` when a job in `todo` state gets a cancel request.
- `aborting` when a job in `doing` state gets a cancel request.
- `aborted` when a job in `aborting` was really aborted.

For more details, have a look at [the documentation page](https://github.com/procrastinate-org/procrastinate/blob/51015b82d2bce623e8a8d24e36293784e8ecca19/docs/howto/advanced/cancellation.md).

> 
> Let me also expand a bit on what my intentions were in regard to cancellation as part of the worker refactoring.
> 
> My idea is to introduce a setting to be able to tell procrastinate how long it should wait for in flight jobs to complete by themselves during the graceful shutdown (the current default is forever).
> Whenever it goes beyond this duration, the worker would cancel the underlying (asyncio) tasks.

This sounds to me more like a timeout instead of a cancellation. But I am fully with you here. A timeout feature would be great, but I understood this issue (or, better to say, my PR) more about a user-initiated cancellation.

> Now, it is a different kettle of fish for synchronous tasks.

That's why I wanted to stay agnostic about synchronous/asynchronous tasks. But as I said, this is for user-initiated cancellation only. For a possible timeout feature, it may make sense to use the `asyncio.CancelledError`.

**Comment by onlyann** [comment]
> That's why I wanted to stay agnostic about synchronous/asynchronous tasks. But as I said, this is for user-initiated cancellation only. For a possible timeout feature, it may make sense to use the asyncio.CancelledError.

Do you think it could not be done with [task.cancel](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel) regardless of it being a worker shutdown or a user initiated explicit job cancellation?

In the case of an async task, it would by default be aborted without any user code. It could also use any of the [shielding from cancellation options ](https://docs.python.org/3/library/asyncio-task.html#shielding-from-cancellation) if it preferred not to.

As for a synchronous task, would it be reasonable to require that task to be changed to async and deal with the cancellation from asyncio as they see fit?

Am I missing something here that requires to introduce an additional API for what is seemingly the same concept?

**Comment by medihack** [comment]
> > That's why I wanted to stay agnostic about synchronous/asynchronous tasks. But as I said, this is for user-initiated cancellation only. For a possible timeout feature, it may make sense to use the asyncio.CancelledError.
> 
> Do you think it could not be done with [task.cancel](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel) regardless of it being a worker shutdown or a user initiated explicit job cancellation?
> 
> In the case of an async task, it would by default be aborted without any user code. It could also use any of the [shielding from cancellation options ](https://docs.python.org/3/library/asyncio-task.html#shielding-from-cancellation) if it preferred not to.
> 
> As for a synchronous task, would it be reasonable to require that task to be changed to async and deal with the cancellation from asyncio as they see fit?
> 
> Am I missing something here that requires to introduce an additional API for what is seemingly the same concept?

From a library user perspective, I find it much easier to reason about calling an API like `context.should_abort()` instead of handling asyncio stuff like CancelledError. Maybe that's more my personal opinion (I am an asyncio user, but not a fan). I also wanted to introduce something that works nearly the same with sync and async code, because Procrastinate (at least at the moment) supports task definitions in both ways.
Would it be possible to use `task.cancel`? Yes, but not with the implementation I did in PR #1081 directly. It could be an additional feature on top of it, but I think it would need listen-notify for that. Maybe combined with an additional option, like `job_manager.cancel_job_by_id(33, raise_async=True)`. But that way, we could have both worlds.

**Comment by ewjoachim** [comment]
> On that note, it begs the question of whether procrastinate wants to keep the burden of handling sync code itself in the future.
If the answer is no, then maybe it is good enough. Otherwise, it sounds like a harder problem to solve.

I can tell you I regret the decision of making Procrastinate support both. I can also tell you if I had to choose and keep one, I'd choose async.
I can also tell you that from the feedback I've seen from actual users, sync is used much much more than async. Also, there are quite a few Django users, and Django isn't 100% async ready yet, even if we're getting there. So the day where we can drop sync support isn't on the horizon, and will probably never be (and I'm probably as sorry as you about this if not more)

**Comment by onlyann** [comment]
> I can tell you I regret the decision of making Procrastinate support both. I can also tell you if I had to choose and keep one, I'd choose async.
I can also tell you that from the feedback I've seen from actual users, sync is used much much more than async. Also, there are quite a few Django users, and Django isn't 100% async ready yet, even if we're getting there. So the day where we can drop sync support isn't on the horizon, and will probably never be (and I'm probably as sorry as you about this if not more)

Fair enough. Possibly this library would have not taken off if it only supported async.

I applaud the constant effort to remain widely compatible.
